src/main/kotlin/org/jetbrains/research/ml/ast/transformations/constantfolding/PyEvaluatorImproved.kt

```diff
@@ -0,0 +1,199 @@
+package org.jetbrains.research.ml.ast.transformations.constantfolding
+
+import com.jetbrains.python.PyTokenTypes
+import com.jetbrains.python.psi.PyBinaryExpression
+import com.jetbrains.python.psi.PyExpression
+import com.jetbrains.python.psi.PyPrefixExpression
+import com.jetbrains.python.psi.impl.PyEvaluator
+import java.math.BigInteger
+import kotlin.reflect.full.declaredFunctions
+import kotlin.reflect.jvm.isAccessible
+
+class PyEvaluatorImproved : PyEvaluator() {
+    private val evaluationResults: MutableMap<PyExpression?, Any?> = mutableMapOf()
+    private val purityResults: MutableMap<PyExpression?, Boolean> = mutableMapOf()
+
+    init {
+        enableResolve(false)
+        setEvaluateKeys(false)
+        setEvaluateCollectionItems(false)
+    }
+
+    override fun evaluate(expression: PyExpression?): Any? =
+        evaluationResults.getOrPut(expression) { evaluateNoCache(expression) }
+
+    private fun evaluateNoCache(expression: PyExpression?): Any? =
+        try {
+            when (expression) {
+                is PyBinaryExpression -> tryEvaluateBinaryHere(expression)
+                is PyPrefixExpression -> tryEvaluatePrefixHere(expression)
+                else -> null
+            } ?: super.evaluate(expression)
+        } catch (_: ArithmeticException) {
+            null
+        }
+
+    private fun evaluateAsPureBoolean(expression: PyExpression?): Boolean? =
+        asBoolean(evaluate(expression))?.takeIf { canBeProvenPure(expression) }
+
+    private fun evaluateAsImpureBoolean(expression: PyExpression?): Boolean? = asBoolean(evaluate(expression))
+
+    private fun canBeProvenPure(expression: PyExpression?): Boolean =
+        purityResults.getOrPut(expression) { canBeProvenPureNoCache(expression) }
+
+    private fun canBeProvenPureNoCache(expression: PyExpression?): Boolean =
+        when (val result = evaluate(expression)) {
+            is Boolean -> true
+            is Number -> true
+            is String -> true
+            is PyExpression -> canBeProvenPure(result)
+            is Collection<*> -> result.all { canBeProvenPure(it as? PyExpression) }
+            is Map<*, *> -> result.all {
+                canBeProvenPure(it.key as? PyExpression) && canBeProvenPure(it.value as? PyExpression)
+            }
+            else -> false
+        }
+
+    private fun tryEvaluateBinaryHere(expression: PyBinaryExpression): Any? {
+        val lhs = expression.leftExpression ?: return null
+        val rhs = expression.rightExpression ?: return null
+        val operator = expression.operator ?: return null
+
+        if (operator in listOf(PyTokenTypes.AND_KEYWORD, PyTokenTypes.OR_KEYWORD)) {
+            evaluateAsPureBoolean(expression.leftExpression)?.let { leftAsPureBool ->
+                if (leftAsPureBool && operator == PyTokenTypes.AND_KEYWORD ||
+                    !leftAsPureBool && operator == PyTokenTypes.OR_KEYWORD
```

Comment:
I think you can create a function and use it here and bellow (send `! leftAsImpureBool as param`) (71-72 lines). What do you think? To be honest, I'm a little confused now by almost the same conditions in two different places

Reply:
I think adding this as a function would be even more confusing :thinking: I've added some comments to clarify what those conditions represent

Reply:
It is also a good way, thanks! My point was that I was a little confused at the sight of two almost identical conditions